src/System.Management.Automation/engine/parser/ast.cs

```diff
@@ -5342,6 +5342,215 @@ internal override AstVisitAction InternalVisit(AstVisitor visitor)
+        /// <param name="chainOperator">The operator used.</param>
+        /// <param name="extent">The extent of the chained statement.</param>
+        public StatementChainAst(
+            IScriptExtent extent,
+            PipelineBaseAst lhsStatement,
+            StatementAst rhsStatement,
+            TokenKind chainOperator)
+            : base(extent)
+        {
+            if (lhsStatement == null)
+            {
+                throw new ArgumentNullException(nameof(lhsStatement));
+            }
+
+            if (rhsStatement == null)
+            {
+                throw new ArgumentNullException(nameof(rhsStatement));
+            }
+
+            if (chainOperator != TokenKind.AndAnd && chainOperator != TokenKind.OrOr)
+            {
+                throw new ArgumentException(nameof(chainOperator));
+            }
+
+            LhsStatement = lhsStatement;
+            RhsStatement = rhsStatement;
+            Operator = chainOperator;
+
+            SetParent(LhsStatement);
+            SetParent(RhsStatement);
+        }
+
+        /// <summary>
+        /// The left hand statement in the chain.
+        /// </summary>
+        public PipelineBaseAst LhsStatement { get; }
+
+        /// <summary>
+        /// The right hand statement in the chain.
+        /// </summary>
+        public StatementAst RhsStatement { get; }
+
+        /// <summary>
+        /// The chaining operator used.
+        /// </summary>
+        public TokenKind Operator { get; }
+
+        /// <summary>
+        /// Create a copy of this Ast.
+        /// </summary>
+        public override Ast Copy()
+        {
+            return new StatementChainAst(Extent, CopyElement(LhsStatement), CopyElement(RhsStatement), Operator);
+        }
+
+        internal override object Accept(ICustomAstVisitor visitor)
+        {
+            if (!(visitor is ICustomAstVisitor2 visitor2))
+            {
+                return null;
+            }
+
+            return visitor2.VisitStatementChain(this);
+        }
+
+        internal override AstVisitAction InternalVisit(AstVisitor visitor)
+        {
+            var visitAction = AstVisitAction.Continue;
+            if (visitor is AstVisitor2 visitor2)
+            {
+                visitAction = visitor2.VisitStatementChain(this);
+
+                if (visitAction != AstVisitAction.Continue)
+                {
+                    return visitAction;
+                }
+            }
+
+            visitAction = LhsStatement.InternalVisit(visitor);
+
+            if (visitAction == AstVisitAction.StopVisit)
+            {
+                return visitAction;
+            }
+
+            return RhsStatement.InternalVisit(visitor);
+        }
+    }
+
+    /// <summary>
+    /// A command-oriented flow-controlled pipeline chain.
+    /// E.g. <c>npm build &amp;&amp; npm test</c> or <c>Get-Content -Raw ./file.txt || "default"</c>.
+    /// </summary>
+    public class PipelineChainAst : PipelineBaseAst
```

Comment:
I'm still not comfortable with having two ASTs for the chain operator ...
A proposal. Add another abstract class in middle of `StatementAst` and `ThrowStatementAst`, `BreakStatementAst`, `ContinueStatementAst`.
```
public abstract ChainableStatementAst : StatementAst { ... }
public abstract class PipelineBaseAst : StatementAst { ... }

public class ThrowStatementAst : ChainableStatementAst  { ... }
public class BreakStatementAst : ChainableStatementAst  { ... }
public class ContinueStatementAst : ChainableStatementAst  { ... }

public class ChainAst : ChainableStatementAst { ... }
```
Then we can have only one `ChainAst` that takes `ChainableStatementAst` for both lhs and rhs. The constructor can have some logic to mark a `ChainAst` is suitable for put in the left-hand side.

Reply:
This proposal is technically a breaking change, as for example `[ThrowStatementAst].BaseType` is not `[StatementAst]` anymore. But this feels to me fall into a grey area and could be acceptable.

Reply:
The reason for the different AST types though isn't about LHS/RHS checking, but to differentiate pipelines from statements when only pipelines are allowed.

So `if (get-item ./dir && get-item ./dir/subdir) { ... }` shouldn't allow for `throw` in it and I don't want it to be possible to construct an AST that the parser would disallow syntactically.

BUT perhaps:

- There should not be chaining in pipeline-only places
- The `throw` stuff is just a bad idea
- It should be a semantic check instead

Anyway, that's the kind of question I was hoping the RFC would have resolved by now...

Reply:
I thought about this some more.
Actually, thinking about the new `switch` expression introduced in C# 8, now the `switch` keyword has two folds of roles: the traditional `switch` statement, and the new `switch` expression.
From that perspective, I feel it might be fine to have two ASTs for the chain operations.

I suggest the following changes to the `PipelineChainAst` and `StatementChainAst`.
The purpose is to make the new AST types indicate the acceptable children accurately, but let me know if you think it doesn't make sense.

```
public abstract ChainableStatementAst : StatementAst { ... }
public class ThrowStatementAst : ChainableStatementAst  { ... }
public class BreakStatementAst : ChainableStatementAst  { ... }
public class ContinueStatementAst : ChainableStatementAst  { ... }

public class StatementChainAst : StatementAst
{
    public StatementChainAst(
            IScriptExtent extent,
            PipelineAst lhsStatement,
            ChainableStatementAst rhsStatement,
            TokenKind chainOperator)
    ...
}

public class PipelineChainAst : PipelineBaseAst
{
    public PipelineChainAst(
            IScriptExtent extent,
            PipelineAst lhsPipeline,
            PipelineAst rhsPipeline,
            TokenKind chainOperator)
    ...
}
```

Reply:
Yeah I like that. I liked your original suggestion too btw, just wanted to solve this bifurcation problem.

Part of me wonders if this is worth doing (trying to specify too much in a grammar or an AST structure is a known trap, compared to just doing semantic checks later and making the grammar simpler). My issue is that PowerShell can construct an AST by reflection and run it. But I suppose semantic checks would still kick in there.

Thoughts?

Reply:
> My issue is that PowerShell can construct an AST by reflection and run it.

If an AST is created by using the constructor and then get the script block from `Ast.GetScriptBlock()`, then we are already guarded. The semantic check will kick in if it hasn't.
If a script block is created using the manually created AST using reflection through the `internal ScirptBlock(IParameterMetadataProvider ast, bool isFilter)` constructor, then today no semantics check action will be taken. But I'm not too worried about this, since there will be a lot more holes if we take reflection attack into account.

I agree with depending on the semantics check. But the option should also depend on how much more work (how much additional grammar) are we looking at.

Reply:
Thinking about this more, I don't like that multiple AST types refer to syntactic elements that visually look the same; it's not clear implementing an AST visitor that you need to implement both `VisitPipelineChain` and `VisitStatementChain` to hit all occurrences of `&&` and `||`.

So I'll write a new commit that puts it all into a `StatementChainAst` and sets a boolean in the constructor to indicate if it must not contain statements